056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,158 @@
+# Kafka Roller KRaft Support
+
+This proposal describes the actions that the KafkaRoller should take when operating 
+against a Strimzi cluster in KRaft mode.
+The proposal describes the checks the KafkaRoller should take, how it should perform 
+those checks and in what order but does not discuss exactly how the KafkaRoller works.
+This proposal is expected to apply to both the current KafkaRoller and a future iteration of the KafkaRoller.
+
+This proposal assumes that liveness/readiness of nodes is as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md) 
+and [PR #8892](https://github.com/strimzi/strimzi-kafka-operator/pull/8892).
+
+
+## Current situation
+
+When operating on a ZooKeeper based cluster the KafkaRoller has the behaviour 
+described below.
+
+### Order to roll nodes
+Roll in order:
+ 1. Unready brokers
+ 2. Ready brokers
+ 3. Current controller
+
+### Triggers
+The following are some of the triggers that roll a broker:
+- Pod or its StrimziPodSet is annotated for manual rolling update
+- Pod is unready
+- Broker's configuration has changed
+
+### Rollability checks
+ 1. Does not force roll a broker performing log recovery
+ 2. Attempts to connect an admin client to a broker and if it can't connect rolls the broker
+ 3. Force rolls a broker if the pod is in stuck state
+ 4. Rolls a broker if the configuration has changed and cannot be updated dynamically
+ 5. Does not roll a broker if doing so would take the in-sync replicas count below `min.insync.replicas`
+
+
+#### Unready pod
+If pod is unready, KafkaRoller awaits its readiness before doing anything. If it hasn't become ready still, it checks whether it is stuck and considers to force roll. A pod is considered stuck if it is in one of following states:
+- `CrashLoopBackOff`
+- `ImagePullBackOff`
+- `ContainerCreating`
+- `Pending` and `Unschedulable`
```

Comment:
Out of curiosity, does the KafkaRoller evaluate how long a pod is "stuck" before rolling? I mean, the `Pending` could be temporary as well as `ContainerCreating` which I am not sure really means it's stuck ... this is a state where a pod always transition to before being ready.

Reply:
@ppatierno KafkaRoller waits for pod's readiness until operational timeout is reached (300000ms by default). After this, if the pod is still in one of these states, it is considered stuck.

Reply:
This is tricky. It might be just a race condition during a pod startup and I think it happens from time to time under some circumstances. But in many other cases, the operator was already waiting for it for 5 minutes. So further delay might not be desired.

Reply:
@tinaselenge this is not clear to me. What do you mean with operational timeout? Is it the reconciliation timeout or something else? Can you also add this part to the stuck pod description?

Reply:
@tinaselenge Can you add the timeout part to the stuck pod description?

Reply:
We've updated the proposal to be more explicit here on what happens when a pod is deemed "stuck". Is there anything else that needs addressing or can we view this thread as resolved?